### PR TITLE
MockitoTestListener must be remove on case test fail on the initialization.

### PR DESCRIPTION
**Pull request by yyvess**
On case of  `MockitoAnnotations.initMocks(target)`  throw any exception `this.target` still null.
As impact to throw a **NullPointerException on all next tests** on the method `DefaultInternalRunner.testFinished`.

What append, the method testFinished is never call on case of any error on the initialization phase as impact that the listener is not cleanup.
On next tests the method testFinished will notify the previous listener that have not been unregistered.

To avoid this issue the mockitoListener must be remove if test fail and never started.

> java.lang.NullPointerException 
at org.mockito.internal.junit.util.TestName.getTestName(TestName.java:15)
	at org.mockito.internal.junit.MismatchReportingTestListener.testFinished(MismatchReportingTestListener.java:33)
	at org.mockito.internal.runners.DefaultInternalRunner$1$1.testFinished(DefaultInternalRunner.java:60)
	at org.junit.runner.notification.SynchronizedRunListener.testFinished(SynchronizedRunListener.java:56)
`

### CODE REVIEW

File: src/test/java/org/mockito/internal/runners/DefaultInternalRunnerTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    };
+
+    @Test
+    public void nominal() throws InitializationError {
+        InternalRunner runner = new DefaultInternalRunner(SuccessTest.class, supplier);
+        runner.run(failIfFailure());
+    }
+
+    @Test
+    public void second_test_must_success() throws InitializationError {
```
Comment (TimvdLippe) [inline src/test/java/org/mockito/internal/runners/DefaultInternalRunnerTest.java]: Please add a more descriptive name here. Probably something like `does_not_throw_NPE_when_setup_failed_after_second_run()`?

File: src/test/java/org/mockito/internal/runners/DefaultInternalRunnerTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        InternalRunner failRunner = new DefaultInternalRunner(TestFailOnInitialization.class, supplier);
+        failRunner.run(new RunNotifier());
+
+        InternalRunner successRunner = new DefaultInternalRunner(SuccessTest.class, supplier);
+        successRunner.run(failIfFailure());
```
Comment (TimvdLippe) [inline src/test/java/org/mockito/internal/runners/DefaultInternalRunnerTest.java]: Can we add assertions for these invocations? Would like to make sure the tests were actually executed as expected.

File: src/test/java/org/mockito/internal/runners/DefaultInternalRunnerTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    private final Supplier<MockitoTestListener> supplier = new Supplier<MockitoTestListener>() {
+        public MockitoTestListener get() {
+            return new NoOpTestListener();
+        }
+    };
+
+    @Test
+    public void nominal() throws InitializationError {
+        InternalRunner runner = new DefaultInternalRunner(SuccessTest.class, supplier);
+        runner.run(failIfFailure());
```
Comment (TimvdLippe) [inline src/test/java/org/mockito/internal/runners/DefaultInternalRunnerTest.java]: Please add an assertion here too.

File: src/test/java/org/mockito/internal/runners/DefaultInternalRunnerTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+public class DefaultInternalRunnerTest {
+
+    private final Supplier<MockitoTestListener> supplier = new Supplier<MockitoTestListener>() {
+        public MockitoTestListener get() {
+            return new NoOpTestListener();
+        }
+    };
+
+    @Test
+    public void nominal() throws InitializationError {
```
Comment (TimvdLippe) [inline src/test/java/org/mockito/internal/runners/DefaultInternalRunnerTest.java]: `does_not_fail_when_tests_succeeds()`

File: src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java [hunk 5]
```diff
@@ -26,44 +26,50 @@
...
-
                 // init annotated mocks before tests
                 MockitoAnnotations.initMocks(target);
-                this.target = target;
                 return super.withBefores(method, target, statement);
             }
 
             public void run(final RunNotifier notifier) {
-                RunListener listener = new RunListener() {
+                final RunListener listener = new RunListener() {
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java]: Nit: not really needed

File: src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java [hunk 6]
```diff
@@ -26,44 +26,50 @@
     public DefaultInternalRunner(Class<?> testClass, final Supplier<MockitoTestListener> listenerSupplier) throws InitializationError {
         runner = new BlockJUnit4ClassRunner(testClass) {
 
-            public Object target;
+            private Object target;
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java]: This might be a breaking change. Let's revert it for now.

File: src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java [hunk 7]
```diff
@@ -26,44 +26,50 @@
...
                     Throwable failure;
+                    boolean started;
+
+                    public void testStarted(Description description) throws Exception {
+                        started = true;
+                    }
 
                     public void testFailure(Failure failure) throws Exception {
                         this.failure = failure.getException();
+                        // If the test have failed before to started remove listener
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java]: Let's make this a bit more clear:
```
// If the test fails during the setup, `testFinished` is never invoked
// Therefore, if we have not started, cleanup the testlistener
```

**Comment by yyvess** [comment]
All remarks have been fixed.

File: src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java [hunk 8]
```diff
@@ -30,40 +30,50 @@ public DefaultInternalRunner(Class<?> testClass, final Supplier<MockitoTestListe
...
+                        // If the test fails during the setup, `testFinished` is never invoked
+                        // Therefore, if we have not started, cleanup the testlistener
+                        if (!started && mockitoTestListener != null) {
+                            Mockito.framework().removeListener(mockitoTestListener);
+                        }
                     }
 
+                    @Override
                     public void testFinished(Description description) throws Exception {
-                        super.testFinished(description);
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java]: Why did you remove this?

Comment (yyvess) [inline src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java]: Because the class RunListener is documented that 
have no implementation.
It more like a interface with defaut empty implementation.

https://github.com/junit-team/junit4/blob/master/src/main/java/org/junit/runner/notification/RunListener.java

Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/runners/DefaultInternalRunner.java]: Ah okay, that seems fine!

**Comment by TimvdLippe** [comment]
Merging this one, seems like a solid fix. Thanks for your work!

**Comment by szczepiq** [comment]
Wow. I was just reviewing changes for potential notable release and I found this PR. Thank you very much for fixing it. The problem was pretty deep in our code, nice job! I made it noteworthy and it will be included in notable release notes when we publish the next release to maven central ([more details](https://github.com/mockito/mockito/wiki/Continuous-Delivery-Details#cd)).

**Comment by yyvess** [comment]
For the storry. This issue have been detected during a migration from Mockito 1.x / PowerMock to Mockito 2.x. 
After fixed small compilation issues, most of tests will failed.
Little bit desperate at first time. 
In reallity most of failures has related to this issue and only few tests requierd a fix to finalize the migration.
Tanks for your API.
